app/celer-g4/GeantDiagnostics.hh

```diff
@@ -0,0 +1,95 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celer-g4/GeantDiagnostics.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "corecel/Assert.hh"
+#include "corecel/io/OutputRegistry.hh"
+#include "accel/SharedParams.hh"
+
+#include "GeantStepDiagnostic.hh"
+
+namespace celeritas
+{
+namespace app
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Diagnostics for Geant4 (i.e., for tracks not offloaded to Celeritas).
+ *
+ * A single instance of this class should be created by the master thread and
+ * shared across all threads.
+ */
+class GeantDiagnostics
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using SPConstParams = std::shared_ptr<SharedParams const>;
+    using SPOutputRegistry = std::shared_ptr<OutputRegistry>;
+    using SPStepDiagnostic = std::shared_ptr<GeantStepDiagnostic>;
+    //!@}
+
+  public:
+    // Construct in an uninitialized state
+    GeantDiagnostics() = default;
+
+    // Construct from shared Celeritas params on the master thread
+    explicit GeantDiagnostics(SPConstParams params);
+
+    // Initialize diagnostics on the master thread
+    inline void Initialize(SPConstParams params);
+
+    // Write (shared) diagnostic output
+    void Finalize();
+
+    // Access the step diagnostic
+    inline SPStepDiagnostic StepDiagnostic() const;
```

Comment:
Return a const reference to the SP to avoid copying it, which uses an atomic (!) and can kill multithread performance